proxy/src/main/java/com/wavefront/agent/ProxyUtil.java

```diff
@@ -0,0 +1,189 @@
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Miscellaneous support methods for running Wavefront proxy.
+ *
+ * @author vasily@wavefront.com
+ */
+abstract class ProxyUtil {
+  protected static final Logger logger = Logger.getLogger("proxy");
+
+  private ProxyUtil() {
+  }
+
+  /**
+   * Gets or creates  proxy id for this machine.
+   *
+   * @param proxyConfig proxy configuration
+   * @return proxy ID
+   */
+  static UUID getOrCreateProxyId(ProxyConfig proxyConfig) {
+    if (proxyConfig.isEphemeral()) {
+      UUID proxyId = UUID.randomUUID(); // don't need to store one
+      logger.info("Ephemeral proxy id created: " + proxyId);
+      return proxyId;
+    } else {
+      return getOrCreateProxyIdFromFile(proxyConfig.getIdFile());
+    }
+  }
+
+  /**
+   * Read or create proxy id for this machine. Reads the UUID from specified file,
+   * or from ~/.dshell/id if idFileName is null.
+   *
+   * @param idFileName file name to read proxy ID from.
+   * @return proxy id
+   */
+  static UUID getOrCreateProxyIdFromFile(@Nullable String idFileName) {
+    File proxyIdFile;
+    UUID proxyId = UUID.randomUUID();
+    if (idFileName != null) {
+      proxyIdFile = new File(idFileName);
+    } else {
+      File userHome = new File(System.getProperty("user.home"));
+      if (!userHome.exists() || !userHome.isDirectory()) {
+        logger.severe("Cannot read from user.home, quitting");
+        System.exit(1);
+      }
+      File configDirectory = new File(userHome, ".dshell");
+      if (configDirectory.exists()) {
+        if (!configDirectory.isDirectory()) {
+          logger.severe(configDirectory + " must be a directory!");
+          System.exit(1);
+        }
+      } else {
+        if (!configDirectory.mkdir()) {
+          logger.severe("Cannot create .dshell directory under " + userHome);
+          System.exit(1);
+        }
+      }
+      proxyIdFile = new File(configDirectory, "id");
+    }
+    if (proxyIdFile.exists()) {
+      if (proxyIdFile.isFile()) {
+        try {
+          proxyId = UUID.fromString(Objects.requireNonNull(Files.asCharSource(proxyIdFile,
+              Charsets.UTF_8).readFirstLine()));
+          logger.info("Proxy Id read from file: " + proxyId);
+        } catch (IllegalArgumentException ex) {
+          logger.severe("Cannot read proxy id from " + proxyIdFile +
+              ", content is malformed");
+          System.exit(1);
+        } catch (IOException e) {
+          logger.log(Level.SEVERE, "Cannot read from " + proxyIdFile, e);
+          System.exit(1);
+        }
+      } else {
+        logger.severe(proxyIdFile + " is not a file!");
+        System.exit(1);
+      }
+    } else {
+      logger.info("Proxy Id created: " + proxyId);
```

Comment:
would this make more sense after the try/catch since each startup would potentially then create a new proxyId (unless the file was manually created, and this is intended as a means to give the user a way to bypass failing permissions??)

Reply:
By default we use ephemeral mode which generates new proxy id at each startup, it just doesn't persist it in a file. A new proxy ID is only valid if it's been associated with a valid token with proper permissions.